ckanext/geocat/harvester.py

```diff
@@ -20,168 +17,152 @@
-                context = {
-                    'model': model,
-                    'session': Session,
-                    'ignore_auth': True
-                }
-                source_dataset = get_action('package_show')(
-                    context, {'id': harvest_job.source_id})
-                self.config['organization'] = source_dataset.get(
-                    'organization').get('name')
-        except GeocatConfigError, e:
+            csw_data = csw_processor.GeocatCatalogueServiceWeb(url=csw_url)
+            gathered_geocat_identifiers = csw_data.get_geocat_id_from_csw()
+        except Exception as e:
             self._save_gather_error(
-                'Config value missing: %s' % str(e),
+                'Unable to get content for URL: %s: %s / %s'
+                % (csw_url, str(e), traceback.format_exc()),
                 harvest_job
             )
             return []
 
-        csw_url = None
-        harvest_obj_ids = []
-        gathered_dataset_identifiers = []
+        existing_dataset_infos = search_utils.get_dataset_infos_for_organization(  # noqa
+            organization_name=self.config['organization'],
+            harvest_source_id=harvest_job.source_id,
+        )
 
-        try:
-            csw_url = harvest_job.source.url.rstrip('/')
-            csw = md.CswHelper(url=csw_url)
-
-            cql = self.config.get('cql', None)
-            if cql is None:
-                cql = "keyword = 'opendata.swiss'"
-
-            log.debug("CQL query: %s" % cql)
-            for record_id in csw.get_id_by_search(cql=cql):
-                harvest_obj = HarvestObject(
-                    guid=record_id,
-                    job=harvest_job
-                )
-                harvest_obj.save()
-                harvest_obj_ids.append(harvest_obj.id)
-                gathered_dataset_identifiers.append('%s@%s' % (
-                    record_id,
-                    self.config['organization']
-                ))
+        gathered_ogdch_identifiers = \
+            [ogdch_map_utils.map_geocat_to_ogdch_identifier(geocat_identifier=geocat_identifier,  # noqa
+                                                  organization_slug=self.config['organization'])  # noqa
+             for geocat_identifier in gathered_geocat_identifiers]
+
+        all_ogdch_identifiers = set(gathered_ogdch_identifiers + existing_dataset_infos.keys())  # noqa
 
-            log.debug('IDs: %r' % harvest_obj_ids)
-        except Exception, e:
+        packages_to_delete = search_utils.get_packages_to_delete(
+            existing_dataset_infos=existing_dataset_infos,
+            gathered_ogdch_identifiers=gathered_ogdch_identifiers,
+        )
+        packages_that_cannot_be_created = search_utils.get_double_packages(
+            existing_dataset_infos=existing_dataset_infos,
+            gathered_ogdch_identifiers=gathered_ogdch_identifiers,
+        )
+        for identifier, info in packages_that_cannot_be_created:
```

Comment:
With your change, we remove the identifiers from `gathered_ogdch_identifiers`, but it's `gathered_geocat_identifiers` that we loop through in order to do the import, so that doesn't actually change anything. We need to remove the geocat identifiers that correspond to the ones from from `gathered_ogdch_identifiers` from `gathered_geocat_identifiers` before mapping and saving.

We could also do this check for each ogdch_identifier after it is calculated in line 138 below, rather than checking all of them at once, and then save the error and move on to the next geocat_identifier in `gathered_geocat_identifiers`.

Reply:
Well spotted and I liked the second approach better, so I implemented it.

Reply:
thanks! :)